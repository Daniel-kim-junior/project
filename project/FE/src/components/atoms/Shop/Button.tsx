import React from "react";
import styled from "@emotion/styled";
import cn from "classnames";

interface ButtonProps {
  children: React.ReactNode;
  flex: string;
  color: string;
  outline: string;
  bgColor: string;
  size: string;
  type: "submit" | "reset" | "button";
  url?: string;
  after?: boolean;
  className: string;
  onClick?: React.MouseEventHandler<HTMLElement>;
}

const Button = ({
  flex = "auto",
  children,
  color = "black",
  outline = "black",
  bgColor = "white",
  size = "normal",
  type = "button",
  after,
  className,
  onClick,
}: ButtonProps) => {
  const classCandiate = [size, className];
  const commonProps = {
    flex,
    children,
    color,
    outline,
    bgColor,
    size,
    onClick,
    type,
    after,
  };
  return (
    <StyledButton {...commonProps} className={cn(classCandiate)}>
      {children}
    </StyledButton>
  );
};

const StyledButton = styled.button<ButtonProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  outline: none;
  flex: ${(props: ButtonProps) => props.flex};
  border: ${(props: ButtonProps) =>
    props.outline === "none" ? "none" : `0.7px solid ${props.outline}`};
  color: ${(props: ButtonProps) => props.color};
  background-color: ${(props: ButtonProps) =>
    props.bgColor === "transparent" ? "transparent" : props.bgColor};
  &.small {
    padding: 7px 7px;
    font-size: 1rem;
  }

  &.normal {
    padding: 10px 10px;
    font-size: 1.2rem;
  }
  &.big {
    padding: 14px 14px;
    font-size: 1.4rem;
  }

  &::after {
    content: "";
    display: ${(props: ButtonProps) => (props.after ? "block" : "none")};
    background-color: #dfe0e5;
    width: 0.1rem;
    height: 1.6rem;
  }
`;

export default Button;
