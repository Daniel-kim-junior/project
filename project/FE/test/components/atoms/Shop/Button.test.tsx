import React from "react";
import Button from "../../../../src/components/atoms/Shop/Button";
import { cleanup } from "@testing-library/react";
import * as TestRenderer from "react-test-renderer";
import { matchers } from "@emotion/jest";

expect.extend(matchers);

describe("Button", () => {
  // explicitly declare type

  let render: TestRenderer.ReactTestRendererJSON | TestRenderer.ReactTestRendererJSON[];
  let dom: TestRenderer.ReactTestRenderer;
  let domRoot: TestRenderer.ReactTestInstance;

  const method = () => {};

  beforeEach(() => {
    render = TestRenderer.create(
      <Button
        flex="auto"
        color="black"
        transparent
        bgColor="black"
        outline="black"
        size="2em"
        type="button"
        className="normal"
        onClick={method}
        after
        children="버튼"
      />
    ).toJSON()!;
    dom = TestRenderer.create(
      <Button
        flex="auto"
        color="black"
        transparent
        bgColor="black"
        outline="black"
        size="2em"
        type="button"
        className="normal"
        onClick={method}
        after
        children="버튼"
      />
    )!;
    domRoot = dom.root!;
  });

  it("Button Component Snapshot", () => {
    // Test first Render
    expect(render).toMatchSnapshot();
  });

  it("Button Component Completely render props", () => {
    expect(render).toMatchInlineSnapshot(`
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  outline: none;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  border: 0.7px solid black;
  color: black;
}

.emotion-0.small {
  padding: 7px 7px;
  font-size: 1rem;
}

.emotion-0.normal {
  padding: 10px 10px;
  font-size: 1.2rem;
}

.emotion-0.big {
  padding: 14px 14px;
  font-size: 1.4rem;
}

.emotion-0::after {
  content: "";
  display: none;
  background-color: #dfe0e5;
  width: 0.1rem;
  height: 1.6rem;
}

<button
  className="2em normal emotion-0"
  color="black"
  onClick={[Function]}
  size="2em"
  type="button"
>
  버튼
</button>
`);
  });

  it("Button should have color", () => {
    expect(render).toHaveStyleRule("color", "black");
  });

  it("Button should have flex", () => {
    expect(render).toHaveStyleRule("flex", "auto");
  });
  it("Button should have prop 버튼", () => {
    expect(domRoot.findByType(Button).props.children).toBe("버튼");
  });
});

afterEach(cleanup);

// 만들어야 될 atom    svg use, span, a, ul li
