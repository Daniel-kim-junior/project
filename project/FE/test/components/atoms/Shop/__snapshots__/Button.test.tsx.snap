// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button Component Snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  outline: none;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  border: 0.7px solid black;
  color: black;
}

.emotion-0.small {
  padding: 7px 7px;
  font-size: 1rem;
}

.emotion-0.normal {
  padding: 10px 10px;
  font-size: 1.2rem;
}

.emotion-0.big {
  padding: 14px 14px;
  font-size: 1.4rem;
}

.emotion-0::after {
  content: "";
  display: none;
  background-color: #dfe0e5;
  width: 0.1rem;
  height: 1.6rem;
}

<button
  className="2em normal emotion-0"
  color="black"
  onClick={[Function]}
  size="2em"
  type="button"
>
  버튼
</button>
`;
